### PR TITLE
ION Metadata Standard signal source

### PR DESCRIPTION
**Pull request by castle055**
This is a signal source implementation that can read XML files that follow ION's GNSS Metadata Standard and properly extract sample streams on demand.

It should cover most common cases and serve as a starting point for further development. All of the infrastructure is present, but some of the more specific bit-wise operations could use more testing.

It is important to note that this implementation is slower than a regular file signal source. This is because ION's metadata standard allows for sub-byte samples, for instance, packing three 5-bit samples inside two bytes (with 1 bit padding). Bit-wise operations are needed to unpack these samples which hurts performance. This is difficult to overcome since different streams may be packed together, each with their own bit format.

I am introducing a dependency on [IonMetadataWorkingGroup/GNSS-Metadata-Standard](https://github.com/IonMetadataWorkingGroup/GNSS-Metadata-Standard) since it implements the parser for the metadata standard.

# Usage:

A metadata file is basically a structured way of presenting a list of sample streams that are somehow stored somewhere. Where and how the bits themselves are found is completely determined within the metadata file. This means that the user does not need to know things like quantization, alignment or packing. The only relevant configuration left for the user is which streams to decode and which to skip.

This is simple enough, as the metadata file identifies each stream with a unique string ID. Thus the user need only provide a comma separated list of stream IDs. These streams will be read, decoded and presented as different output channels, thus supporting multi-channel receivers. GNSS-SDR can determine item types and sizes automatically from the metadata file.

Example configuration:
```ini
SignalSource.implementation=ION_GSMS_Signal_Source  
SignalSource.metadata_filename=data/capture0_L1.sdrx  
SignalSource.streams=L1
```

And for testing purposes, here is an example metadata file that properly describes the data file found in [My first position fix](https://gnss-sdr.org/my-first-fix/).
```xml
<?xml version="1.0" encoding="ISO-8859-1"?>  
<metadata xmlns="http://www.ion.org/standards/sdrwg/schema/metadata.xsd">  
    <system id="test-system">  
        <freqbase format="MHz"> 4.0000000000000000e+00 </freqbase>  
    </system>  
    <band id="L1">  
        <centerfreq format="GHz"> 1.5754200000000000e+00 </centerfreq>  
        <translatedfreq format="kHz"> 0.0000000000000000e+00 </translatedfreq>  
        <bandwidth format="MHz"> 2.4000000000000000e+01 </bandwidth>  
    </band>  
    <file>  
        <url>2013_04_04_GNSS_SIGNAL_at_CTTC_SPAIN.dat</url>  
        <timestamp>04-Apr-2013 00:00:00</timestamp>  
        <lane id="L1"/>  
    </file>  
    <lane id="L1">  
        <bandsrc idband="L1" idsrc="none"/>  
        <system id="test-system"/>  
        <block>  
            <cycles>0</cycles>  
            <chunk>  
                <sizeword>2</sizeword>  
                <countwords>2</countwords>  
                <endian>Little</endian>  
                <padding>None</padding>  
                <wordshift>Left</wordshift>  
                <lump>  
                    <stream id="L1">  
                        <ratefactor>1</ratefactor>  
                        <quantization>16</quantization>  
                        <packedbits>32</packedbits>  
                        <alignment>Right</alignment>  
                        <shift>Left</shift>  
                        <format>IQ</format>  
                        <encoding>TC</encoding>  
                        <band id="L1"/>  
                    </stream>  
                </lump>  
            </chunk>  
        </block>  
    </lane>  
</metadata>
```

### CODE REVIEW

File: src/algorithms/signal_source/libs/ion_gsms.cc
Comment (jwmelto) [inline src/algorithms/signal_source/libs/ion_gsms.cc:89]: This seems a bit obtuse. Consider:
```suggestion
    items_produced_.clear();
    items_produced_.resize(output_items.size());
```
Note that `resize()` initializes all _new_ items to `T()` so by making the vector length 0 before resizing it, you ensure all the valuse are default initialized.

File: src/algorithms/signal_source/libs/ion_gsms.cc
Comment (jwmelto) [inline src/algorithms/signal_source/libs/ion_gsms.cc:99]: there is nothing intuitive about this loop (or the `work()` function at all, for that matter). Please provide some comments explaining what is happening so the next maintainer can follow along.

File: src/algorithms/signal_source/libs/ion_gsms.h
Comment (jwmelto) [inline src/algorithms/signal_source/libs/ion_gsms.h:37]: why do you define a signature with unused parameters? There should be nothing to be compatible with that requires a generic API here. In fact, as I understand the paradigm, the Adapter should translate from the Configuration to construct the blocks; the blocks should be completely insulated from the config mechanism.

File: src/algorithms/signal_source/libs/ion_gsms.h
Comment (jwmelto) [inline src/algorithms/signal_source/libs/ion_gsms.h:71]: include directives should be at the top of the file

**Review by jwmelto — CHANGES_REQUESTED** [review]
_DISCLAIMER: I have no authority on this project. I'm just an interested contributor. Take my comments for what they are worth or ignore them._

At a minimum, ensure that this new dependency is not needed if it isn't wanted (and not wanting it should be the default case). Stated differently, make sure explicit action is required to enable this.

I can't tell how much of what I don't like about this API is inherited from the package you imported, and how much was personal coding style. This new capability may be very useful, but the code seems almost needlessly complex in many places. It doesn't feel quite ready for prime time, to me.

**Comment by carlesfernandez** [comment]
I have addressed some of the mentioned issues here: https://github.com/carlesfernandez/gnss-sdr/tree/castle055-ft/ion-metadata-standard-signal-source

@castle055 please consider merging that branch into yours, and keep working on the code addressing other issues raised by @jwmelto (thanks for the review!).

Changes are:

* Fixes for C++ standards older than 20. Avoid C++20-specific lambda templates.
* Include missing includes.
* Make the source opt-in instead of opt-out: define ENABLE_ION to OFF by default. This is required for packaging, since this source requires a non-standard dependency.
* Create a CMake target for the ION dependency for consistency.

I've noticed that the receiver does not stop when it reaches the end of the raw samples file. Please check out how other adapters solve this. I've also noticed that the receiver becomes terribly slow. Could you tell what's the bottleneck operation here?

File: CMakeLists.txt
Comment (castle055) [reply CMakeLists.txt:1434]: Honestly, I wanted to use their library since it is their standard and you would think it would be in a better state. But the latest commit on their `devel` branch is "Fix Memory Leak", it's 4 years old and hasn't been merged yet.

I will fix a commit, since there are no tags apart from the branches themselves.

File: src/algorithms/signal_source/adapters/ion_gsms_signal_source.cc
Comment (castle055) [reply src/algorithms/signal_source/adapters/ion_gsms_signal_source.cc:25]: This is the logging and flags mechanism that GNSS_SDR uses. Taken right from `file_source_base.cc`. But I may have been too excitedly adding it to header files too instead of just source files.

I saw @carlesfernandez took care of that in his branch. I will merge it into my branch and make sure it's only included where it's needed.

File: src/algorithms/signal_source/adapters/ion_gsms_signal_source.cc
Comment (castle055) [reply src/algorithms/signal_source/adapters/ion_gsms_signal_source.cc:58]: As far as I can tell, it is quite common for signal sources to do this, the `file_source_base` sets default values for every path in its configuration, including the sample file (which is aptly set as "../data/example_capture.data", so I went with something similar).

As for adding a test, if that metadata file does not exist, the program will crash with an error.

File: src/algorithms/signal_source/adapters/ion_gsms_signal_source.cc
Comment (castle055) [reply src/algorithms/signal_source/adapters/ion_gsms_signal_source.cc:60]: You are right, `metadata_file_` does not need to be a member field. I could initialize `metadata_` with the configuration lookup call directly.

File: src/algorithms/signal_source/adapters/ion_gsms_signal_source.cc
Comment (castle055) [reply src/algorithms/signal_source/adapters/ion_gsms_signal_source.cc:89]: It is, but I could add a comment explaining where that `copy_blocks_` vector comes from. You can check that this vector is initialized in the constructor.

I could still add the assertion out of an abundance of caution, just in case this code is modified in the future by someone other than me.

File: src/algorithms/signal_source/adapters/ion_gsms_signal_source.cc
Comment (castle055) [reply src/algorithms/signal_source/adapters/ion_gsms_signal_source.cc:120]: good catch

File: src/algorithms/signal_source/libs/ion_gsms.h
Comment (castle055) [reply src/algorithms/signal_source/libs/ion_gsms.h:37]: I didn't realize about that separation of configuration concerns between adapters and gnuradio_blocks. I figured it would be nice to have the configuration available in there just in case I needed it.

I will remove it.

File: src/algorithms/signal_source/libs/ion_gsms.h
Comment (castle055) [reply src/algorithms/signal_source/libs/ion_gsms.h:71]: You spotted the duck tape.

I completely missed this, I should have never committed it. Will fix!

**Comment by castle055** [comment]
Thank you @jwmelto for the detailed review, and thank you @carlesfernandez for the changes. I have merged them into my branch, I will spend this afternoon fixing the other issues.

### Regarding performance:

Yes, this signal source is slow, but there is a reason for that. ION's metadata standard allows for a myriad of packing structures, I will explain with an example:

Let's say we have three sample streams multiplexed into the same file. The first and second streams encoded as 3 bit complex samples, and the third as 2 bit complex samples. These can all be packed withing a 2 byte word. It would look something like this:

```
[ 1 1 1 0 0 0 1 1 1 0 0 0 1 1 0 0 ] <- this is read from the file as a short or as two bytes
  \___/ \___/ \___/ \___/ \_/ \_/
   I1    Q1    I2    Q2    I3  Q3
```

This is just an example as the standard allows for pretty much any format to be described by the metadata file.

The streams need to be unpacked and separated into different channels, some might not even be needed and thus can be skipped. The problem here is that we have to extract each individual sample one by one using bit-wise operations. Compare that to the regular file signal source, which just copies a long chunk of the sample file into its output buffer, and the reason for the slow down is obvious.

There may be ways of making it faster, but I couldn't think of any.

**Comment by carlesfernandez** [comment]
Thanks @castle055 for your work. Another round of suggested changes: https://github.com/carlesfernandez/gnss-sdr/tree/castle055-ft/ion-metadata-standard-signal-source

**Comment by castle055** [comment]
Thank you for the fixes! I have merged them into my branch.

File: CMakeLists.txt
Comment (carlesfernandez) [reply CMakeLists.txt]: Now all files related to this source are conditional to `ENABLE_ION`, which is set to `OFF`  by default.

File: CMakeLists.txt
Comment (carlesfernandez) [reply CMakeLists.txt]: Fixed: pointing to the current last commit in the `master` branch.

File: src/algorithms/signal_source/libs/ion_gsms.cc
Comment (carlesfernandez) [reply src/algorithms/signal_source/libs/ion_gsms.cc]: Fixed

File: src/algorithms/signal_source/libs/ion_gsms.h
Comment (carlesfernandez) [reply src/algorithms/signal_source/libs/ion_gsms.h]: Fixed

File: src/algorithms/signal_source/adapters/ion_gsms_signal_source.cc
Comment (carlesfernandez) [reply src/algorithms/signal_source/adapters/ion_gsms_signal_source.cc]: Fixed

File: src/algorithms/signal_source/libs/CMakeLists.txt
Comment (carlesfernandez) [reply src/algorithms/signal_source/libs/CMakeLists.txt]: Fixed

File: src/algorithms/signal_source/libs/ion_gsms.cc
Comment (carlesfernandez) [reply src/algorithms/signal_source/libs/ion_gsms.cc]: Fixed

File: src/algorithms/signal_source/libs/ion_gsms.cc
Comment (carlesfernandez) [reply src/algorithms/signal_source/libs/ion_gsms.cc]: Added comments

**Comment by carlesfernandez** [comment]
Hi @castle055 and @jwmelto,

I believe we've resolved the majority of the major issues, and this is now ready to be merged. Thank you very much for your valuable contributions. I'm adding a few final commits to address some of the remaining open comments. Further exposure to this source should help identify any additional limitations.

I've written some documentation here: https://gnss-sdr.org/docs/sp-blocks/signal-source/#implementation-ion_gsms_signal_source (should be online a few minutes after the merge). Please feel free to comment on that or suggest changes.

**Comment by castle055** [comment]
I agree, @carlesfernandez , I am comfortable with the current state of the code. I have just merged your [next](https://github.com/carlesfernandez/gnss-sdr/tree/next) branch to include the CI fix as well as the other changes.

Again, thank you for your work and the quick responses!

**Comment by carlesfernandez** [comment]
Merged

**Pull request closed without merging**
